issue #2189

muhammadbilal19 opened the pull request:
Add Pagination code for IncomingPhoneNumber API and UI. Please review and let me know with your feedback.

muhammadbilal19 commented:
Thanks @gvagenas  i will complete requested changes.

jaimecasero commented on restcomm/restcomm.application/src/main/webapp/WEB-INF/sql/incoming-phone-numbers.xml:
```diff
@@ -68,7 +68,21 @@
     <if test="phoneNumber != null">
       AND "n"."phone_number" like #{phoneNumber}
     </if>
+    LIMIT #{limit} OFFSET #{offset}
```
pagination filtering done at DB level should provide better performance, good!!

jaimecasero commented on restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/IncomingPhoneNumbersEndpoint.java:
```diff
@@ -261,14 +268,62 @@ protected Response getAvailableCountries(final String accountSid, final MediaTyp
+
+        int limit = Integer.parseInt(pageSize);
+        int offset = (page == "0") ? 0
+                : (((Integer.parseInt(page) - 1) * Integer.parseInt(pageSize)) + Integer.parseInt(pageSize));
```
can we reduce the number of times we convert incoming queryString params into proper type for validation?? For performance and error handling point of view good be much better...

muhammadbilal19 replied:
Thanks @jaimecasero very nice point, changes are now available :) Feel free to comment if you have any other point

muhammadbilal19 commented:
@gvagenas  your requested changes are now available. 

- Mybatis mapper is added for Mysql
- Tested with Mysql DB
- Added test cases in IncomingPhoneNumbersEndpointTest

Feel free to review it now and let me know if you have any other comments. Due to some mistake i added few extra commits.

scottbarstow commented:
@gvagenas Are we close on this do you think?

gvagenas commented:
@scottbarstow @malihanaveed checking now

muhammadbilal19 commented:
@gvagenas i tested it with my local mysql database. Is live DB is some different database ?

gvagenas commented:
@muhammadbilal19 if you tested with a mysql server its fine

muhammadbilal19 commented:
@ammendonca i will try to fix these issues today

muhammadbilal19 commented:
@ammendonca items per page button and sorting issue is fixed for this PR. Please have a look and let me know your comments

ammendonca commented:
Thanks @muhammadbilal19, the mentioned issues are now fixed. A (new?) minor issue seems to be present tho, having 12 numbers, with the list with 10 per page, on the page 2 it shows "3 to 4 of 12 Numbers". Not sure this was present before and I haven't noticed.

![image](https://user-images.githubusercontent.com/1737954/27145394-94fb61b0-512d-11e7-8de6-4f5f835eee8a.png)

scottbarstow commented:
@*ammendonca @gvagenas *can we get this deployed on cloud asap?

On Wed, Jun 14, 2017 at 1:16 PM, Alexandre Mendonça <
notifications@github.com> wrote:

> *@ammendonca* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RestComm/Restcomm-Connect/pull/2211#pullrequestreview-44086846>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAtjo-i2kyjec4NGcaM916b45h7oXD2Fks5sEBWIgaJpZM4N03TK>
> .
>



-- 
scottbarstow.com
@scottbarstow
+1-919-349-9473 (m)

muhammadbilal19 commented:
@ammendonca Thanks for the update. I just tried to reproduce that minor issue which you mentioned but i am unable to reproduced it. Not sure why it is happening on your side.
<img width="1134" alt="screen shot 2017-06-15 at 8 38 15 am" src="https://user-images.githubusercontent.com/8377490/27153695-8faaf828-51a6-11e7-90f7-d9035c532485.png">

ammendonca commented:
@muhammadbilal19 I just checked and you need to sort by numbers ascending for it to show. Actually looking at the sorted list, it doesn't seem to be properly sorted:

![image](https://user-images.githubusercontent.com/1737954/27155127-99c5e874-514f-11e7-91c0-3faa27d6770a.png)

@scottbarstow for me it's fine, I've approved the PR already. The issue I've found about record numbering is minor and not blocking.. the fact that the list isn't correctly sorted might be more relevant in cases where we have long list of numbers.

muhammadbilal19 commented:
@ammendonca yes i am able to see that issue now. I am not a really good UI developer and tester so bear with me :) I guess it would be good if we create another issue for sorting issue and once @gvagenas completed the review and approved that PR merge pagination changes only.
If we need to fix sorting issue with this PR i am working on it and will commit it once it is fixed.
@scottbarstow @deruelle thoughts??

scottbarstow commented:
I'd rather not do two releases so if we can fix it while @gvagenas reviews
then just merge to the existing PR prior to merge to master that's ideal.

Right now I'd roll it in with this issue if @ammendonca is ok with it. If
it's a stopper, then we fix now. His call

On Wed, Jun 14, 2017 at 8:19 PM, Muhammad Bilal <notifications@github.com>
wrote:

> @ammendonca <https://github.com/ammendonca> yes i am able to see that
> issue now. I am not a really good UI developer and tester so bear with me
> :) I guess it would be good if we create another issue for sorting issue
> and once @gvagenas <https://github.com/gvagenas> completed the review and
> approved that PR merge pagination changes only.
> If we need to fix sorting issue with this PR i am working on it and will
> commit it once it is fixed.
> @scottbarstow <https://github.com/scottbarstow> @deruelle
> <https://github.com/deruelle> thoughts??
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RestComm/Restcomm-Connect/pull/2211#issuecomment-308593620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAtjow4SY_4ox9P3xEe967gaXL8HfldMks5sEHiOgaJpZM4N03TK>
> .
>



-- 
scottbarstow.com
@scottbarstow
+1-919-349-9473 (m)

ammendonca commented:
@muhammadbilal19 Just tested and you can fix the sort issue by replacing `sort` in `<tr class="number-row" ng-repeat="pn in filtered = (numbersList | filter:query| orderBy:sort:reverse)">` by `predicate` which is the actual variable containing the field to sort by.

As per the "3 to 4 of 12 Numbers", the issue is with `page_size` returned when the `Reverse=true` is used. With `Reverse=false` the page_size (even for the last page, which has 2 entries) is **10**. But if we pass `Reverse=true` the page_size returned is **2**. That page_size is also used to generate the returned `start` and `end` params which are used to show that text.

Hope it helps!

@scottbarstow I'm OK to go with current status, but just provided the solution to the most critical UI issue, so that should likely make it in.

ammendonca commented on restcomm/restcomm.ui/src/main/webapp/modules/numbers-incoming.html:
```diff
@@ -22,14 +22,16 @@
             </thead>
             <tbody>
-            <tr class="number-row" ng-repeat="pn in filtered = (numbersList | filter:query)">
+            <tr class="number-row" ng-repeat="pn in filtered = (numbersList | filter:query| orderBy:sort:reverse)">
```
Replace `sort` by `predicate` which is where the field to sort is stored at (from https://github.com/RestComm/Restcomm-Connect/pull/2211/files#diff-b0cba56bc32531ce8ec4f5174dcd4fe6R97)

muhammadbilal19 commented:
Thanks @ammendonca for writing the solution I committed my latest changes  i made some other changes as well so please have and look and let me know if looks good to you now.Sorting issue and other minor issue both are now fixed

gvagenas commented:
Great work @muhammadbilal19, thanks!

gvagenas commented:
@muhammadbilal19 can you please create a new pull request for the same issue since there was a problem during the merge and I had to revert ?

muhammadbilal19 commented:
@gvagenas  sure i am creating new PR give me few minutes

gvagenas commented:
Great, thanks @muhammadbilal19 . Make sure you merge `master` into your branch to have the latest changes
